### DISCUSSION TITLE
Versioning and tags for consequent release

### DISCUSSION THREAD

**Discussion by gcarreno** [wp-xyz/corona#24] [Ideas]
Hey Werner,

You should begin thinking about a version. And you should then do your first tag so I can add the binaries for that release.

In my Object Pascal apps I use all 4 version parts: Major.Minor.Revision.Build Ex: 1.2.3.4.
You can decide how many or how few of those you want and if you want only numbers, or numbers and letters.
Only numbers is what I'm suggesting.
Number and letters is what you can find on [fpcupdeluxe](https://github.com/LongDirtyAnimAlf/fpcupdeluxe/releases): 1.8.2j

Some like to not include the 'v' before the numbers when creating the tag, others include it.

If you have any questions about tagging, please feel free to ask. I'll make my best effort to answer your questions.

We'll tackle this after the GitHub actions.

Cheers,
Gus

**Comment by wp-xyz** [comment]
Versioning means that I must activate the version information in the project settings? Or is this something which is completely independent of the project-internal version? Anyway, I prefer the four-number format with preceding "v": "v1.0.0.317"

**Reply by gcarreno** [reply]
Hey Werner (@wp-xyz),

Ermmm, simple question with a rather complex answer counting the times I've said hummm in my head while trying to formulate a coherent answer.
But let me try and untangle this ball of wool...

I'm not gonna talk about the Lazarus option for project version. I think you got that covered.

What you do on GIT is more of a convention than something that connects physically or digitally.

So first you come up with your convention, as you already did: `v[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+`
On a side note [semver](https://semver.org/) (Semantic Versioning) is the broadly accepted in the industry.

Then right before you `git tag` the current stage of you project, you make sure that all the files/project files contain that version.

I, myself, have a very complicated way of doing it:
- On the project I turn version on
- I have the same numbers on one of the units as constants
  - I have a string with the full version, no v. I'll add that with a `Format()`
  - I have 4 integer constants with Major, Minor, Revision and Build
- I use those constants on various Captions around the project
- I `grep` that source file to extract the version to create `MyApp-v[version].tgz`, etc

But there is no connection apart from your convention and how convoluted you want to do your versions.

I have it quite convoluted:
1. Update the Project version on `CTRL+Shift+F11`
2. Update the String constant
3. Update the integer constants
4. Build the whole project: `Shift+F9`
5. Call the script that creates the builds and packs the binaries for various platforms
6. Call `git tag -a v0.1.0.1 -m "Release v0.1.0.1"`
7. Call `git push --tags`
8. On GitHub create a release with the usual blah, blah about changes
9. Upload the binaries
10. Relax and have a cuppa...

Hope that answers your question, which to put it in simpler terms, YOU determine how simple or complex your versioning system is, but alas, no connection from Lazarus to GIT apart from the one you make with a convention.

Cheers,
Gus
